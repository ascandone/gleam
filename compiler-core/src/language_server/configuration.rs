use serde::Deserialize;
use std::sync::{Arc, RwLock};

pub type SharedConfig = Arc<RwLock<Configuration>>;

#[derive(Debug, Default, Clone, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct Configuration {
    pub inlay_hints: InlayHintsConfig,
}

#[derive(Debug, Default, Clone, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct InlayHintsConfig {
    /// Whether to show type inlay hints of multiline pipelines
    pub pipelines: bool,
}
