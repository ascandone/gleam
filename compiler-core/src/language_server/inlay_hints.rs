use lsp_types::{InlayHint, InlayHintKind, InlayHintLabel};

use crate::{
    ast::{
        PipelineAssignmentKind, SrcSpan, TypedExpr, TypedModule, TypedPipelineAssignment,
        visit::Visit,
    },
    line_numbers::LineNumbers,
    type_::{self, Type},
};

use super::src_offset_to_lsp_position;

struct InlayHintsVisitor<'a> {
    module_names: &'a type_::printer::Names,
    current_declaration_printer: type_::printer::Printer<'a>,

    hints: Vec<InlayHint>,
    line_numbers: &'a LineNumbers,
}

impl<'a> InlayHintsVisitor<'a> {
    fn new(
        line_numbers: &'a LineNumbers,
        module_names: &'a type_::printer::Names,
    ) -> InlayHintsVisitor<'a> {
        InlayHintsVisitor {
            module_names,
            current_declaration_printer: type_::printer::Printer::new(module_names),
            hints: vec![],
            line_numbers,
        }
    }
}

fn default_inlay_hint(line_numbers: &LineNumbers, offset: u32, label: String) -> InlayHint {
    let position = src_offset_to_lsp_position(offset, line_numbers);

    InlayHint {
        position,
        label: InlayHintLabel::String(label),
        kind: Some(InlayHintKind::TYPE),
        text_edits: None,
        tooltip: None,
        padding_left: Some(true),
        padding_right: None,
        data: None,
    }
}

impl InlayHintsVisitor<'_> {
    pub fn push_binding_annotation(&mut self, type_: &Type, span: &SrcSpan) {
        let label = format!(": {}", self.current_declaration_printer.print_type(type_));

        let mut hint = default_inlay_hint(self.line_numbers, span.end, label);
        hint.padding_left = Some(false);

        self.hints.push(hint);
    }

    pub fn push_return_annotation(&mut self, type_: &Type, span: &SrcSpan) {
        let label = format!("-> {}", self.current_declaration_printer.print_type(type_));

        let hint = default_inlay_hint(self.line_numbers, span.end, label);

        self.hints.push(hint);
    }
}

impl<'ast> Visit<'ast> for InlayHintsVisitor<'_> {
    fn visit_typed_function(&mut self, fun: &'ast crate::ast::TypedFunction) {
        // This must be reset on every statement
        self.current_declaration_printer = type_::printer::Printer::new(self.module_names);

        for arg in &fun.arguments {
            if arg.annotation.is_none() {
                self.push_binding_annotation(&arg.type_, &arg.location);
            }
        }

        if fun.return_annotation.is_none() {
            self.push_return_annotation(&fun.return_type, &fun.location);
        }

        for st in &fun.body {
            self.visit_typed_statement(st);
        }
    }

    fn visit_typed_expr_fn(
        &mut self,
        _location: &'ast SrcSpan,
        type_: &'ast std::sync::Arc<Type>,
        kind: &'ast crate::ast::FunctionLiteralKind,
        args: &'ast [crate::ast::TypedArg],
        body: &'ast vec1::Vec1<crate::ast::TypedStatement>,
        return_annotation: &'ast Option<crate::ast::TypeAst>,
    ) {
        if let crate::ast::FunctionLiteralKind::Anonymous { head } = kind {
            for arg in args {
                if arg.annotation.is_none() {
                    self.push_binding_annotation(&arg.type_, &arg.location);
                }
            }

            if return_annotation.is_none() {
                if let Some((_args, ret_type)) = type_.fn_types() {
                    self.push_return_annotation(&ret_type, head);
                }
            }
        }

        for st in body {
            self.visit_typed_statement(st);
        }
    }

    fn visit_typed_expr_pipeline(
        &mut self,
        _location: &'ast SrcSpan,
        first_value: &'ast TypedPipelineAssignment,
        assignments: &'ast [(TypedPipelineAssignment, PipelineAssignmentKind)],
        finally: &'ast TypedExpr,
        _finally_kind: &'ast PipelineAssignmentKind,
    ) {
        let mut prev_hint: Option<(u32, Option<InlayHint>)> = None;

        let assigments_values =
            std::iter::once(first_value).chain(assignments.iter().map(|p| &p.0));

        for assign in assigments_values {
            let this_line: u32 = self
                .line_numbers
                .line_and_column_number(assign.location.end)
                .line;

            if let Some((prev_line, prev_hint)) = prev_hint {
                if prev_line != this_line {
                    if let Some(prev_hint) = prev_hint {
                        self.hints.push(prev_hint);
                    }
                }
            };

            let this_hint = default_inlay_hint(
                self.line_numbers,
                assign.location.end,
                self.current_declaration_printer
                    .print_type(assign.type_().as_ref())
                    .to_string(),
            );

            prev_hint = Some((
                this_line,
                if assign.value.is_simple_lit() {
                    None
                } else {
                    Some(this_hint)
                },
            ));

            self.visit_typed_expr(&assign.value);
        }

        if let Some((prev_line, prev_hint)) = prev_hint {
            let this_line = self
                .line_numbers
                .line_and_column_number(finally.location().end)
                .line;
            if this_line != prev_line {
                if let Some(prev_hint) = prev_hint {
                    self.hints.push(prev_hint);
                }
                let hint = default_inlay_hint(
                    self.line_numbers,
                    finally.location().end,
                    self.current_declaration_printer
                        .print_type(finally.type_().as_ref())
                        .to_string(),
                );
                self.hints.push(hint);
            }
        }

        self.visit_typed_expr(finally);
    }
}

pub fn get_inlay_hints(typed_module: TypedModule, line_numbers: &LineNumbers) -> Vec<InlayHint> {
    let mut visitor = InlayHintsVisitor::new(line_numbers, &typed_module.names);
    visitor.visit_typed_module(&typed_module);
    visitor.hints
}
