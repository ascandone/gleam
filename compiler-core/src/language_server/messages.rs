use crate::language_server::configuration::Configuration;
use camino::Utf8PathBuf;
use lsp::{
    notification::{DidChangeWatchedFiles, DidOpenTextDocument},
    request::GotoDefinition,
};
use lsp_types::{
    self as lsp,
    notification::{DidChangeTextDocument, DidCloseTextDocument, DidSaveTextDocument},
    request::{
        CodeActionRequest, Completion, DocumentSymbolRequest, Formatting, GotoTypeDefinition,
        HoverRequest, InlayHintRequest, PrepareRenameRequest, Rename, SignatureHelpRequest,
    },
};
use std::{collections::HashMap, time::Duration};

#[derive(Debug)]
pub enum Message {
    Request(lsp_server::RequestId, Request),
    Response(Response),
    Notification(Notification),
}

#[derive(Debug)]
pub enum Request {
    Format(lsp::DocumentFormattingParams),
    Hover(lsp::HoverParams),
    GoToDefinition(lsp::GotoDefinitionParams),
    GoToTypeDefinition(lsp::GotoDefinitionParams),
    Completion(lsp::CompletionParams),
    CodeAction(lsp::CodeActionParams),
    SignatureHelp(lsp::SignatureHelpParams),
    DocumentSymbol(lsp::DocumentSymbolParams),
    ShowInlayHints(lsp::InlayHintParams),
}

#[derive(Debug)]
pub enum Response {
    Configuration(Configuration),
    PrepareRename(lsp::TextDocumentPositionParams),
    Rename(lsp::RenameParams),
}

impl Request {
    fn extract(request: lsp_server::Request) -> Option<Message> {
        let id = request.id.clone();
        match request.method.as_str() {
            "textDocument/formatting" => {
                let params = cast_request::<Formatting>(request);
                Some(Message::Request(id, Request::Format(params)))
            }
            "textDocument/hover" => {
                let params = cast_request::<HoverRequest>(request);
                Some(Message::Request(id, Request::Hover(params)))
            }
            "textDocument/definition" => {
                let params = cast_request::<GotoDefinition>(request);
                Some(Message::Request(id, Request::GoToDefinition(params)))
            }
            "textDocument/completion" => {
                let params = cast_request::<Completion>(request);
                Some(Message::Request(id, Request::Completion(params)))
            }
            "textDocument/codeAction" => {
                let params = cast_request::<CodeActionRequest>(request);
                Some(Message::Request(id, Request::CodeAction(params)))
            }
            "textDocument/signatureHelp" => {
                let params = cast_request::<SignatureHelpRequest>(request);
                Some(Message::Request(id, Request::SignatureHelp(params)))
            }
            "textDocument/documentSymbol" => {
                let params = cast_request::<DocumentSymbolRequest>(request);
                Some(Message::Request(id, Request::DocumentSymbol(params)))
            }
            "textDocument/inlayHint" => {
                let params = cast_request::<InlayHintRequest>(request);
                Some(Message::Request(id, Request::ShowInlayHints(params)))
            }
            "textDocument/rename" => {
                let params = cast_request::<Rename>(request);
                Some(Message::Request(id, Request::Rename(params)))
            }
            "textDocument/prepareRename" => {
                let params = cast_request::<PrepareRenameRequest>(request);
                Some(Message::Request(id, Request::PrepareRename(params)))
            }
            "textDocument/typeDefinition" => {
                let params = cast_request::<GotoTypeDefinition>(request);
                Some(Message::Request(id, Request::GoToTypeDefinition(params)))
            }
            _ => None,
        }
    }
}

#[derive(Debug)]
pub enum Notification {
    /// A Gleam file has been modified in memory, and the new text is provided.
    SourceFileChangedInMemory { path: Utf8PathBuf, text: String },
    /// A Gleam file has been saved or closed in the editor.
    SourceFileMatchesDisc { path: Utf8PathBuf },
    /// gleam.toml has changed.
    ConfigFileChanged { path: Utf8PathBuf },
    /// The user edited a client config option
    ConfigChanged,
    /// It's time to compile all open projects.
    CompilePlease,
}

impl Notification {
    fn extract(notification: lsp_server::Notification) -> Option<Message> {
        match notification.method.as_str() {
            "textDocument/didOpen" => {
                let params = cast_notification::<DidOpenTextDocument>(notification);
                let notification = Notification::SourceFileChangedInMemory {
                    path: super::path(&params.text_document.uri),
                    text: params.text_document.text,
                };
                Some(Message::Notification(notification))
            }
            "textDocument/didChange" => {
                let params = cast_notification::<DidChangeTextDocument>(notification);
                let notification = Notification::SourceFileChangedInMemory {
                    path: super::path(&params.text_document.uri),
                    text: params.content_changes.into_iter().last()?.text,
                };
                Some(Message::Notification(notification))
            }

            "textDocument/didSave" => {
                let params = cast_notification::<DidSaveTextDocument>(notification);
                let notification = Notification::SourceFileMatchesDisc {
                    path: super::path(&params.text_document.uri),
                };
                Some(Message::Notification(notification))
            }
            "textDocument/didClose" => {
                let params = cast_notification::<DidCloseTextDocument>(notification);
                let notification = Notification::SourceFileMatchesDisc {
                    path: super::path(&params.text_document.uri),
                };
                Some(Message::Notification(notification))
            }

            "workspace/didChangeWatchedFiles" => {
                let params = cast_notification::<DidChangeWatchedFiles>(notification);
                let notification = Notification::ConfigFileChanged {
                    path: super::path(&params.changes.into_iter().last()?.uri),
                };
                Some(Message::Notification(notification))
            }

            "workspace/didChangeConfiguration" => {
                Some(Message::Notification(Notification::ConfigChanged))
            }

            _ => None,
        }
    }
}

pub enum Next {
    MorePlease,
    Handle(Vec<Message>),
    Stop,
}

/// The message buffer pulls messages from the client until one of the following
/// happens:
/// - A shutdown request is received.
/// - A short pause in messages is detected, indicating the programmer has
///   stopped typing for a moment and would benefit from feedback.
/// - A request type message is received, which requires an immediate response.
#[derive(Debug)]
pub struct MessageBuffer {
    messages: Vec<Message>,
    next_request_id: i32,
    response_handlers: HashMap<lsp_server::RequestId, ResponseHandler>,
}

impl MessageBuffer {
    pub fn new() -> Self {
        Self {
            messages: Vec::new(),
            next_request_id: 1,
            response_handlers: Default::default(),
        }
    }

    pub fn receive(&mut self, conn: &lsp_server::Connection) -> Next {
        let pause = Duration::from_millis(100);

        // If the buffer is empty, wait indefinitely for the first message.
        // If the buffer is not empty, wait for a short time to see if more messages are
        // coming before processing the ones we have.
        let message = if self.messages.is_empty() {
            Some(conn.receiver.recv().expect("Receiving LSP message"))
        } else {
            conn.receiver.recv_timeout(pause).ok()
        };

        // If have have not received a message then it means there is a pause in the
        // messages from the client, implying the programmer has stopped typing. Process
        // the currently enqueued messages.
        let message = match message {
            Some(message) => message,
            None => {
                // A compile please message it added in the instance of this
                // pause of activity so that the client gets feedback on the
                // state of the code as it is now.
                self.push_compile_please_message();
                return Next::Handle(self.take_messages());
            }
        };

        match message {
            lsp_server::Message::Request(r) if self.shutdown(conn, &r) => Next::Stop,
            lsp_server::Message::Request(r) => self.request(r),
            lsp_server::Message::Response(r) => self.response(r),
            lsp_server::Message::Notification(n) => self.notification(n),
        }
    }

    fn request(&mut self, r: lsp_server::Request) -> Next {
        let Some(message) = Request::extract(r) else {
            return Next::MorePlease;
        };

        // Compile the code prior to attempting to process the response, to
        // ensure that the response is based on the latest code.
        self.push_compile_please_message();
        self.messages.push(message);
        Next::Handle(self.take_messages())
    }

    fn notification(&mut self, n: lsp_server::Notification) -> Next {
        // A new notification telling us that an edit has been made, or
        // something along those lines.
        if let Some(message) = Notification::extract(n) {
            self.messages.push(message);
        }
        // Ask for more messages (or a pause), at which point we'll start processing.
        Next::MorePlease
    }

    pub fn make_request(
        &mut self,
        method: impl Into<String>,
        params: impl serde::Serialize,
        handler: Option<ResponseHandler>,
    ) -> lsp_server::Request {
        let id = self.next_request_id;
        self.next_request_id += 1;
        let request = lsp_server::Request {
            id: id.into(),
            method: method.into(),
            params: serde_json::value::to_value(params).expect("serialisation should never fail"),
        };

        if let Some(handler) = handler {
            _ = self.response_handlers.insert(id.into(), handler);
        }

        request
    }

    fn configuration_update_received(&mut self, result: serde_json::Value) -> Next {
        let parsed_update_items: Result<(Configuration,), _> = serde_json::from_value(result);
        let Ok((parsed_config,)) = parsed_update_items else {
            return Next::MorePlease;
        };

        let message = Message::Response(Response::Configuration(parsed_config));
        self.messages.push(message);

        Next::Handle(self.take_messages())
    }

    fn handle_response(&mut self, handler: ResponseHandler, result: serde_json::Value) -> Next {
        match handler {
            ResponseHandler::UpdateConfiguration => self.configuration_update_received(result),
        }
    }

    fn response(&mut self, response: lsp_server::Response) -> Next {
        if let Some(handler) = self.response_handlers.remove(&response.id) {
            if let Some(result) = response.result {
                return self.handle_response(handler, result);
            }
        }

        // We do not use or expect responses from the client currently.
        Next::MorePlease
    }

    /// Add a `CompilePlease` message which will prompt the engine to compile
    /// the projects.
    ///
    fn push_compile_please_message(&mut self) {
        let message = Notification::CompilePlease;
        let value = Message::Notification(message);
        self.messages.push(value);
    }

    fn take_messages(&mut self) -> Vec<Message> {
        std::mem::take(&mut self.messages)
    }

    fn shutdown(
        &mut self,
        connection: &lsp_server::Connection,
        request: &lsp_server::Request,
    ) -> bool {
        connection.handle_shutdown(request).expect("LSP shutdown")
    }
}

fn cast_request<R>(request: lsp_server::Request) -> R::Params
where
    R: lsp::request::Request,
    R::Params: serde::de::DeserializeOwned,
{
    let (_, params) = request.extract(R::METHOD).expect("cast request");
    params
}

fn cast_notification<N>(notification: lsp_server::Notification) -> N::Params
where
    N: lsp::notification::Notification,
    N::Params: serde::de::DeserializeOwned,
{
    notification
        .extract::<N::Params>(N::METHOD)
        .expect("cast notification")
}

#[derive(Debug)]
pub enum ResponseHandler {
    UpdateConfiguration,
}
