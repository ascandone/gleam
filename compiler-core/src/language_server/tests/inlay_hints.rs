use crate::language_server::{
    configuration::{Configuration, InlayHintsConfig},
    tests::{LanguageServerTestIO, TestProject, setup_engine},
};
use lsp_types::{InlayHintParams, Position, Range};

#[test]
fn no_hints_when_same_line() {
    let src = r#"
    fn identity(x) {
      x
    }

    fn ret_str(_x) {
      "abc"
    }

    pub fn example_pipe() {
      0 |> ret_str() |> identity()
    }
"#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn no_hints_when_value_is_literal() {
    let src = r#"
    pub fn ret_str(f1) {
      "abc"
      |> f1()
    }

    pub fn ret_int(f2) {
      42
      |> f2()
    }

    pub fn ret_float(f3) {
      42.2
      |> f3()
    }

    pub fn ret_bit_array(f4) {
      <<1, 2>>
      |> f4()
    }
"#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn show_many_hints() {
    let src = r#"
          const int_val = 0

          fn identity(x) {
            x
          }

          fn ret_str(_x) {
            "abc"
          }

          pub fn example_pipe() {
            int_val
            |> ret_str()
            |> identity()
          }
      "#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn hints_nested_in_case_block() {
    let src = r#"
          const int_val = 0

          fn identity(x) {
            x
          }

          fn main(a) {
            case a {
              _ -> {
                  int_val
                  |> identity()
              }
            }
          }
      "#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn hints_nested_for_apply_fn_let() {
    let src = r#"
          const int_val = 0

          fn identity(x) {
            x
          }

          fn main() {
            let f = identity(fn() {
              int_val
              |> identity()
            })
          }
      "#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn hints_in_use() {
    let src = r#"
          const int_val = 0

          fn identity(x) {
            x
          }

          fn main(f) {
            use a <- f()
            int_val
            |> identity()
          }
      "#;

    let hints = pipeline_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn show_hints_in_params() {
    let src = r#"
      fn example_fn(
        flex_type_arg,
        b: do_not_show_this,
      ) { 0 }
      "#;

    let hints = parameter_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn show_hints_in_return() {
    let src = r#"
      fn example_fn() { 0 }
      "#;

    let hints = return_hints(src);
    insta::assert_snapshot!(hints);
}

#[test]
fn show_correct_type_names_in_functions() {
    let src = r#"
      fn complex(
        x, //: b
        y: rigid_type_var, //: rigid_type_var
      ) { //-> fn(a, rigid_type_var) -> #(b, rigid_type_var, a)
        fn(
          z, //: a
        ) { //-> #(b, rigid_type_var, a)
          #(x, y, z)
        }
      }
      "#;

    let hints = inlay_hints_for_config(
        src,
        InlayHintsConfig {
            function_parameter_types: true,
            function_return_types: true,
            ..Default::default()
        },
    );
    insta::assert_snapshot!(hints);
}

#[test]
fn do_not_show_hints_by_default() {
    let src = r#"
          const int_val = 0

          fn identity(x) {
            x
          }

          fn ret_str(_x) {
            "abc"
          }

          pub fn example_pipe() {
            int_val
            |> ret_str()
            |> identity()
          }
      "#;

    let hints = inlay_hints_for_config(src, InlayHintsConfig::default());
    insta::assert_snapshot!(hints);
}

fn pipeline_hints(src: &str) -> String {
    inlay_hints_for_config(
        src,
        InlayHintsConfig {
            pipelines: true,
            ..Default::default()
        },
    )
}

fn parameter_hints(src: &str) -> String {
    inlay_hints_for_config(
        src,
        InlayHintsConfig {
            function_parameter_types: true,
            ..Default::default()
        },
    )
}

fn return_hints(src: &str) -> String {
    inlay_hints_for_config(
        src,
        InlayHintsConfig {
            function_return_types: true,
            ..Default::default()
        },
    )
}

fn inlay_hints_for_config(src: &str, inlay_hints_config: InlayHintsConfig) -> String {
    let io = LanguageServerTestIO::new();
    let mut engine = setup_engine(&io);
    {
        let mut config = engine.user_config.write().expect("cannot write config");
        *config = Configuration {
            inlay_hints: inlay_hints_config,
            ..Default::default()
        };
    }

    _ = io.src_module("app", src);
    let response = engine.compile_please();
    assert!(response.result.is_ok());

    let params = InlayHintParams {
        text_document: TestProject::build_path(),
        work_done_progress_params: Default::default(),
        range: Range::new(
            Position::new(0, 0),
            Position::new(
                src.lines().count() as u32,
                src.lines().last().unwrap_or_default().len() as u32,
            ),
        ),
    };

    let hints = engine
        .inlay_hints(params)
        .result
        .expect("inlay hint request should not fail");

    let stringified = serde_json::to_string_pretty(&hints).expect("json pprint should not fail");

    stringified
}
