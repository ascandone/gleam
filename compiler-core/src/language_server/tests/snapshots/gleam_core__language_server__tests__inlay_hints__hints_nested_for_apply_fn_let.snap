---
source: compiler-core/src/language_server/tests/inlay_hints.rs
expression: hints
---
[
  {
    "position": {
      "line": 9,
      "character": 21
    },
    "label": "Int",
    "kind": 1,
    "paddingLeft": true
  },
  {
    "position": {
      "line": 10,
      "character": 27
    },
    "label": "Int",
    "kind": 1,
    "paddingLeft": true
  }
]
